xarray/core/dataset.py

```diff
@@ -1212,7 +1212,18 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         coords: dict[Hashable, Variable] = {}
         # preserve ordering
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
+            var_dims = set(self._variables[k].dims)
+            if k in self._indexes:
+                try:
+                    self._indexes[k].validate_dataarray_coord(
+                        k, self._variables[k], needed_dims
+                    )
+                    coords[k] = self._variables[k]
+                except ValueError:
+                    # failback to strict DataArray model check (index may be dropped later)
+                    if var_dims <= needed_dims:
+                        coords[k] = self._variables[k]
```

Comment:
It is a little weird how the fallback logic here doesn't depend on the index, but is called inside the loop.

Reply:
I'm not sure to understand? Maybe you've seen `if k in self._indexes` above instead of `for k in self._indexes` above?

Reply:
oops, indeed, I misread this!